### PR TITLE
16.0 reconversion

### PR DESCRIPTION
**Pull request by stephansainleger**

### CODE REVIEW

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 1]
```diff
@@ -13,6 +13,9 @@ class AccountInvoice(models.Model):
     credit_request_ids = fields.One2many(
         "credit.request", "invoice_id", string="Credit Requests"
     )
+    debit_request_ids = fields.One2many(
+        "debit.request", "debit_move_id", string="Credit Requests"
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: typo: the description is wrong

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 2]
```diff
@@ -60,6 +69,19 @@ def _compute_has_numeric_lcc_products(self):
             if lcc_numeric_products:
                 self.has_numeric_lcc_products = True
 
+    def _compute_debit_request_ids(self):
+        for move in self:
+            move.debit_request_ids = self.env["debit.request"].search(["|", ("debit_move_id", "=", move.id), ("commission_move_id", "=", move.id)])
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: there are no commission_move_id yet

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 3]
```diff
@@ -60,6 +69,19 @@ def _compute_has_numeric_lcc_products(self):
...
+
+    @api.depends("debit_request_ids", "credit_request_ids")
+    def _compute_digital_currency_invoice_type(self):
+        if self.debit_request_ids:
+            self.digital_currency_invoice_type = "debit"
+        elif self.credit_request_ids:
+            self.digital_currency_invoice_type = "credit"
+        else:
+            self.digital_currency_invoice_type = "none"
+
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: so if there are both debit request and credit requests, it is considered a debit ... The only place where it is used in current code could do without this (`account_invoice.xml`) using `('credit_request_ids', '=', '[]')`.

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 4]
```diff
@@ -69,8 +91,23 @@ def _invoice_paid_hook(self):
                     continue
                 # Set the state in "pending" to launch the top up process
                 request.write({"state": "pending"})
+        for invoice in self.filtered(lambda move: move.is_invoice() and move.is_purchase_document() and move.has_numeric_lcc_products):
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: Could factorize with previous loop, and have one if on : is it a sale or a purchase ? It'll be more efficient

File: lcc_lokavaluto_app_connection/models/debit_request.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    def is_ready_to_invoice(self, raise_error=False):
+        self.ensure_one()
+        if self.amount <= 0 and raise_error:
+            raise UserError("Amount must be superior to zero.")
+        if not self.wallet_id and raise_error:
+            raise UserError("The wallet is missing.")
+        if not self.transaction_id and raise_error:
+            raise UserError("The transaction ID is missing.")
+        return (self.amount > 0) and self.wallet_id and self.transaction_id
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/debit_request.py]: You are duplicating the tests for no reasons... You could raise always errors in this function, and call it with try/except when you don't want the exceptions.

File: lcc_lokavaluto_app_connection/models/debit_request.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                raise UserError(
+                    "You can't delete a cancelled debit request. Please archive it instead."
+                )
+            elif request.state == "paid":
+                raise UserError(
+                    "You can't delete a paid debit request. Please archive it instead."
+                )
+
+            if request.debit_move_id and request.debit_move_id.state == "draft":
+                request.debit_move_id.unlink()
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/debit_request.py]: you duplicate some tests, and some lines seem unreachable (can you have a state cancelled or paid with a connected non-draft invoice ?). In some case even, it seem it seem it could accept unlink with a `debit_move_id` (but not draft) connected that it'll leave untouched. Currently, I'm not sure if all this is done on purpose.

File: lcc_lokavaluto_app_connection/models/debit_request.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            if not request.debit_move_id:
+                continue
+            if request.debit_move_state == "draft":
+                request.state = "received"
+            if request.debit_move_state == "posted":
+                request.state = "invoiced"
+            if request.debit_move_state == "paid":
+                request.state = "paid"
+            if request.debit_move_id.state == "cancelled":
+                request.state = "cancelled"
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/debit_request.py]: these could be `elif`

File: lcc_lokavaluto_app_connection/models/debit_request.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        if not self.transaction_id and raise_error:
+            raise UserError("The transaction ID is missing.")
+        return (self.amount > 0) and self.wallet_id and self.transaction_id
+
+    #############################
+    ## INVOICE CREATION PROCESSES
+    #############################
+    
+    def create_invoices(self):
+        self.create_debit_invoices()
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/debit_request.py:188]: why this indirection ?

**Review by vaab — COMMENTED** [review]
First commit comments...

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 9]
```diff
@@ -13,8 +13,8 @@ class AccountInvoice(models.Model):
     credit_request_ids = fields.One2many(
         "credit.request", "invoice_id", string="Credit Requests"
     )
-    debit_request_ids = fields.One2many(
-        "debit.request", "debit_move_id", string="Credit Requests"
+    debit_request_ids = fields.Many2many(
+        "debit.request", compute="_compute_debit_request_ids", string="Debit Requests"
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: Shouldn't it still stay a One2Many ?

File: lcc_lokavaluto_app_connection/models/company.py [hunk 10]
```diff
@@ -1,4 +1,4 @@
-from odoo import models, fields
+from odoo import models, fields, api
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py]: not clear why this import is needed in this commit

File: lcc_lokavaluto_app_connection/models/debit_request.py [hunk 11]
```diff
@@ -75,15 +86,44 @@ def unlink(self):
...
+                # If any of the invoices is in Draft, the request is still considered as "received"
+                request.state = self._convert_status("draft")
+                continue
+
+            if request.debit_move_state == "posted" or request.commission_move_state == "posted":
+                request.state = self._convert_status("posted")
+                continue
+
+            if request.debit_move_state == "paid" or request.commission_move_state == "paid":
+                request.state = self._convert_status("paid")
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/debit_request.py]: I'm a little lost into the intent here, and fear that it's a little too brittle.

File: lcc_lokavaluto_app_connection/models/debit_request.py [hunk 12]
```diff
@@ -127,4 +168,33 @@ def _get_debit_invoice_line_values(self):
             'quantity': self.amount,
             'price_unit': product_id.standard_price,
         }
+        return invoice_line_values
+
+    def create_commission_invoices(self):
+        for request in self:
+            self.is_ready_to_invoice(raise_error=True)
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/debit_request.py]: this test is done twice

File: lcc_lokavaluto_app_connection/models/res_partner_backend.py [hunk 13]
```diff
@@ -1,8 +1,10 @@
 from odoo import models, fields, api
 from odoo.exceptions import UserError
-
+from odoo.tools.safe_eval import safe_eval
 import logging
 
+
+
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/res_partner_backend.py]: unecessary, some pass with ruff in incrementally can be applied at the end if you wish

File: lcc_lokavaluto_app_connection/models/res_partner_backend.py [hunk 14]
```diff
@@ -95,4 +97,16 @@ def get_wallet_balance(self):
...
-        return res
\ No newline at end of file
+        return res
+        
+    def get_wallet_commission_rule(self):
+        self.ensure_one()
+        rules = self.env["commission.rule"].search([("active", "=", True)])
+        for rule in rules:
+            # Get all the wallet matching the rule
+            wallets = self.search(safe_eval(rule.wallet_domain))
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/res_partner_backend.py]: I'm not sure to see how, without providing a context, this `safe_eval` is useful...

File: lcc_lokavaluto_app_connection/models/res_partner_backend.py [hunk 15]
```diff
@@ -95,4 +97,16 @@ def get_wallet_balance(self):
...
+            # Check if current wallet (self) is in the matching wallets
+            if wallets.filtered(lambda x: x.id == self.id):
+                # First rule matched is returned
+                return rule
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/res_partner_backend.py]: only the first rule will match... first in which order ? the default order as you didn't specify any `_order`. Is that what you want ? This computation will also be quite costly written like that. You should add the `('id', '=', self.id)` to the search at least and avoid doing the test in python.

**Review by vaab — COMMENTED** [review]
Review of second commit

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 16]
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
...
                 # Set the state in "paid" and update the global request status
-                request.write({"commission_move_state": "paid"})
-                request.compute_state()
-        for invoice in self.filtered(lambda move: move.is_invoice() and move.is_purchase_document() and move.has_numeric_lcc_products):
-            for request in invoice.debit_request_ids:
-                # Set the state in "paid" and update the global request status
-                request.write({"debit_move_state": "paid"})
+                if invoice.is_sale_document():
+                    state_value = {"commission_move_state": "paid"}
+                if invoice.is_purchase_document():
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: elif

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 17]
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
...
+        for invoice in self:
+            if not invoice.is_invoice():
+                continue
+
             for request in invoice.debit_request_ids:
-                # Set the state in "posted" and update the global request status
-                request.write({"debit_move_state": "posted"})
+                if invoice.is_sale_document():
+                    state_value = {"commission_move_state": "posted"}
+                if invoice.is_purchase_document():
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: elif

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 18]
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
...
-                request.compute_state()
-        for invoice in self.filtered(lambda move: move.is_invoice() and move.is_purchase_document() and move.has_numeric_lcc_products):
-            for request in invoice.debit_request_ids:
-                # Set the state in "paid" and update the global request status
-                request.write({"debit_move_state": "paid"})
+                if invoice.is_sale_document():
+                    state_value = {"commission_move_state": "paid"}
+                if invoice.is_purchase_document():
+                    state_value = {"debit_move_state": "paid"}
+                request.write(state_value)
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: if none of the 2 condition above matches, it can fail with an error or worse, write the state of the last element seen

File: lcc_lokavaluto_app_connection/models/account_invoice.py [hunk 19]
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
...
+                continue
+
             for request in invoice.debit_request_ids:
-                # Set the state in "posted" and update the global request status
-                request.write({"debit_move_state": "posted"})
+                if invoice.is_sale_document():
+                    state_value = {"commission_move_state": "posted"}
+                if invoice.is_purchase_document():
+                    state_value = {"debit_move_state": "posted"}
+                request.write(state_value)
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/account_invoice.py]: if none of the 2 condition above matches, it can fail with an error or worse, write the state of the last element seen.
Note that if you are sure that if it can't happen, then make an if/else rather than 2 ifs... and best if you have any doubt, is to raise an clear error.

File: lcc_lokavaluto_app_connection/models/credit_request.py [hunk 20]
```diff
@@ -58,7 +58,7 @@ def write(self, vals):
         for request in self:
             if request.state == "pending":
                 # The top up has been paid, the credit process can start
-                if request.partner_id.company_id.activate_automatic_topup:
+                if self.env.user.company_id.activate_automatic_topup:
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/credit_request.py:59]: are we sure to want to check who is actively calling this `write` ? Not real sure to understand why this change.

**Review by vaab — COMMENTED** [review]
Several typos in the commit message of the third commit.

File: lcc_lokavaluto_app_connection/models/company.py [hunk 21]
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
     monujo_ios_app_url = fields.Char(string="Monujo iOS app URL")
     activate_automatic_topup = fields.Boolean("Activate Automatic Topup")
     commission_product_id = fields.Many2one("product.product", string="Commission Product")
+
+
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py]: pep8 requires only one blank line here ;-)

File: lcc_lokavaluto_app_connection/models/company.py [hunk 22]
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
...
+    def _cron_import_new_digital_currency_debit_requests(self):
+        """Create the debit requests in Odoo for all the debit
+        transactions performed since the last check."""
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py:28]: pep 257 strongly recommend one-line summary, and if multi-line, one-line summary, blank line then any content.

File: lcc_lokavaluto_app_connection/models/company.py [hunk 23]
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
...
+            values = self._build_debit_request_values(transaction)
+            if not values:
+                continue
+            request = self.env["debit.request"].create(values)
+            _logger.info("Debit request created for wallet %s on transaction %s" % (request.wallet_id,transaction.get("tx_id")))
+
+
+    def _retrieve_last_debit_transactions(self):
+        """TO OVERIDE in digital currency backend dedicated add-ons
+        A list of transactions (dictionnary) is expected, with the following data:
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py]: dictionary (only one n).

File: lcc_lokavaluto_app_connection/models/company.py [hunk 24]
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
...
+        # Extract last transactions received in the Safe Wallet
+        transactions = self._retrieve_last_debit_transactions()
+
+        # For each transaction received:
+        # - retrieve the transaction id
+        # - check if any debit request exists with this transaction id
+        # - if yes, do nothing
+        # - if no, create a new debit request
+        for transaction in transactions:
+            tx_id = transaction.get("tx_id", "")
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py]: hmm, why silence this error in the data ? A log at least should be important. And we can probably `continue` also.

File: lcc_lokavaluto_app_connection/models/company.py [hunk 25]
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
...
+        - sender: the Odoo name of the wallet concerned by the debit request,
+        - amount: the amount debited from the wallet,
+        - tx_id: the transaction ID in the digital currency backend
+        - tx_timestamp: the timestamp of the transaction
+        """
+        return []
+
+
+    def _build_debit_request_values(self, transaction):
+        wallet_ids = self.env["res.partner.backend"].search([("name", "=", transaction.get("sender"))])
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py]: have you made sure that the names have a constraint of unicity on wallets ?

File: lcc_lokavaluto_app_connection/models/company.py [hunk 26]
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
...
+            return {}
+        elif len(wallet_ids) > 1:
+            _logger.error("Too many wallets found for debit transaction %s, debit request creation CANCELLED" % transaction.get("tx_id"))
+            return {}
+        values = {
+            "active" : True,
+            "wallet_id" : wallet_ids[0].id,
+            "amount" : transaction.get("amount"),
+            "transaction_id" : transaction.get("tx_id"),
+            "transaction_timestamp": transaction.get("tx_timestamp"),
```
Comment (vaab) [inline lcc_lokavaluto_app_connection/models/company.py]: I'm concerned that you didn't check if any of these values are actually set. As it is the responsability of a plugin, it could be nice to check the content and make a nice helpful message of what was forgotten or what went wrong instead of writing bogus values in database and wait for a crash later one.

**Review by vaab — COMMENTED** [review]
a typo (overridden : 2 r, 2 d) and strange wrapping in comment of commit number 4 ! ;-)

File: lcc_comchain_base/models/company.py [hunk 27]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                        "amount": f"{string_amount[0:-2]}.{string_amount[-2:]}",
+                        "tx_id": tx.hash,
+                        "tx_timestamp": datetime.fromtimestamp(int(full_tx.receivedat["value"])),
+                    })
+
+            company_id.last_block_checked_nb = last_block_id
+        except Exception as e:
+            _logger.error(tools.format_last_exception())
+
+        return res
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: res might be incompletely filled with some transaction of the current block. You should probably reraise the exception ? You probably have a good reason why you didn't.

File: lcc_comchain_base/models/company.py [hunk 28]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                    if not tx.currency:
+                        continue # this transaction is not linked with a currency
+                    if not tx.currency.name.lower() == company_id.comchain_currency_name.lower():
+                        continue # this transaction is about another smart contract
+                    full_tx = tx.full_tx
+                    if not full_tx.is_cc_transaction:
+                        continue # this transaction is not a digital currency transfer
+                    if not full_tx.direction == 1:
+                        continue # this transaction is not a reconversion
+                    _logger.info("Transactions %s retrieved" % tx.hash)
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: typo: transaction~s~

File: lcc_comchain_base/models/company.py [hunk 29]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                    sender_adress = full_tx.add2.lstrip("0x")
+                    matched_wallet = self.env["res.partner.backend"].search([("comchain_id", "ilike", sender_adress)])
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: why `ilike` ?

File: lcc_comchain_base/models/company.py [hunk 30]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                        continue # this transaction is not linked with a currency
+                    if not tx.currency.name.lower() == company_id.comchain_currency_name.lower():
+                        continue # this transaction is about another smart contract
+                    full_tx = tx.full_tx
+                    if not full_tx.is_cc_transaction:
+                        continue # this transaction is not a digital currency transfer
+                    if not full_tx.direction == 1:
+                        continue # this transaction is not a reconversion
+                    _logger.info("Transactions %s retrieved" % tx.hash)
+                    sender_adress = full_tx.add2.lstrip("0x")
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: typo: address (with 2 d)

File: lcc_comchain_base/models/company.py [hunk 31]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                    matched_wallet = self.env["res.partner.backend"].search([("comchain_id", "ilike", sender_adress)])
+                    if len(matched_wallet) == 0:
+                        continue # this transactions comes from a wallet no registered in current Odoo database
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: typo: `not registered` and this transaction~s~

File: lcc_comchain_base/models/company.py [hunk 32]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                    string_amount = str(full_tx.sent)
+                    res.append({
+                        "sender": "comchain:%s" % sender_adress,
+                        "amount": f"{string_amount[0:-2]}.{string_amount[-2:]}",
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: Warning here, this is not guaranteed to work, if full_tx.sent = '1' you are screwed ;-)... you need to `.zfill(2)`

File: lcc_comchain_base/models/company.py [hunk 33]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                        "tx_id": tx.hash,
+                        "tx_timestamp": datetime.fromtimestamp(int(full_tx.receivedat["value"])),
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: This value doesn't mean anything. If your version of pyc3l supports it, try `full_tx.received_at` (which is a datetime already). Otherwise you'll also probably have a wrong datetime (it is an utc timestamp, and you should store timezone aware datetime...)

File: lcc_comchain_base/models/company.py [hunk 34]
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
...
+                    })
+
+            company_id.last_block_checked_nb = last_block_id
+        except Exception as e:
+            _logger.error(tools.format_last_exception())
```
Comment (vaab) [inline lcc_comchain_base/models/company.py]: For me that's probably a bad idea. There are many exception that are meant to break totally the flow of odoo python process. My problem is that you are using a 'except Exception' which is too wide, OR that you don't re-raise it.

**Review by vaab — CHANGES_REQUESTED** [review]
some important issue here.

File: lcc_lokavaluto_app_connection/models/account_invoice.py (hunk 1 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/account_invoice.py]: Fixed in an following commit

File: lcc_lokavaluto_app_connection/models/account_invoice.py (hunk 3 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/account_invoice.py]: fixed in https://github.com/Lokavaluto/lokavaluto-addons/pull/301/commits/16d9cddff8c86287d6c4ce3cade8ac8d266a1550

File: lcc_lokavaluto_app_connection/models/account_invoice.py (hunk 4 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/account_invoice.py]: Fixed in an following commit

File: lcc_lokavaluto_app_connection/models/debit_request.py (hunk 5 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/debit_request.py]: solved, raise_error attribute removed, usage of try/except when needed.

File: lcc_lokavaluto_app_connection/models/debit_request.py (hunk 7 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/debit_request.py]: fixed in a following commit

File: lcc_lokavaluto_app_connection/models/debit_request.py (hunk 8 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/debit_request.py:188]: completed later with another function... it is a Back To The Future design ^^'

File: lcc_lokavaluto_app_connection/models/company.py (hunk 10 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/company.py]: removed

File: lcc_lokavaluto_app_connection/models/res_partner_backend.py (hunk 15 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/res_partner_backend.py]: the idea is that as soon as one rule match, we consider that the reconversion is allowed, whatever the rule is.
I have modified the code to make it less costly.

File: lcc_lokavaluto_app_connection/models/res_partner_backend.py (hunk 14 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/res_partner_backend.py]: well I don't know how it works exactly, but without the safe_eval() it doesn't work ^^'

File: lcc_lokavaluto_app_connection/models/company.py (hunk 24 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/company.py]: add a check to directly continue before performing the search on debit.requests in the following line.
add warning log too.

File: lcc_lokavaluto_app_connection/models/company.py (hunk 25 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/company.py]: no unicity constraint for now, but we log an error if several wallets are found.

File: lcc_lokavaluto_app_connection/models/company.py (hunk 26 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/company.py]: this values are filled by function retrieve_last_debit_transactions() defined in lcc_comchain_base and lcc_cyclos_base add-ons. This function check that these values exist.

File: lcc_comchain_base/models/company.py (hunk 27 above)
Comment (stephansainleger) [reply lcc_comchain_base/models/company.py]: fixed: we use a tmp list `comchain_res` to fill the retrieved transaction, and this `comchain_res` is added to `res` only with all the process went smoothly.

File: lcc_comchain_base/models/company.py (hunk 34 above)
Comment (stephansainleger) [reply lcc_comchain_base/models/company.py]: As I am clearly not a error managment expert yet, I will need your help on that point :-)

File: lcc_comchain_base/models/company.py (hunk 33 above)
Comment (stephansainleger) [reply lcc_comchain_base/models/company.py]: fixed by using directly full_tx.received_at

File: lcc_lokavaluto_app_connection/models/account_invoice.py (hunk 9 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/account_invoice.py]: If my understanding is good, if `debit_request_ids` is a one2many, this means that there should be a many2one field that link the debit request to the invoice.
the issue is that here we have 2 many2one that must be checked : `debit_move_id` and `commission_move_id` : according to the type of invoice (sale or purchase), we should consider one or the other.

File: lcc_lokavaluto_app_connection/models/debit_request.py (hunk 6 above)
Comment (stephansainleger) [reply lcc_lokavaluto_app_connection/models/debit_request.py]: fixed
